lib/sudo-cli/src/lib.rs

```diff
@@ -244,220 +75,268 @@ pub struct SudoOptions {
-        // instead.
-        let mut external_args = opts.external_args.iter().rev();
-        // Then we will add a separator manually so we can reprocess everything by calling this
-        // function recursively. To do this, we need to figure out where this separator actually
-        // goes.
-        let mut index = None;
-
-        for (i, last_arg) in vec_args.iter().enumerate().rev() {
-            // If the arg was an environment variable declaration, then we must put the separator
-            // before this arg.
-            if vec_env_vars[i] {
-                continue;
-            }
-            // If the last external arg is the same as this arg, we know that the separator must go
-            // before this arg.
-            if let Some(last_external) = external_args.next() {
-                if last_arg == last_external {
-                    continue;
+    /// parse command line arguments from the environment and handle errors
+    pub fn parse() -> SudoOptions {
+        match Self::try_parse_from(std::env::args()) {
+            Ok(options) => {
+                if options.help {
+                    eprintln!("{HELP_MSG}");
+                    std::process::exit(1);
                 }
-            }
-            // Otherwise, this arg is already a non-external arg.
-            index = Some(i + 1);
-            break;
-        }
 
-        if let Some(mut index) = index {
-            // There is one last catch. In this command `sudo FOO=1 env BAR=2 ls`, `FOO=1` is not
-            // part of the external command. So we need to ignore the first environment variable
-            // declarations before putting the separator.
-            while vec_env_vars[index] {
-                index += 1;
+                options
+            }
+            Err(e) => {
+                eprintln!("{e}\n{USAGE_MSG}");
+                std::process::exit(1);
             }
-            // Then we insert the separator and parse the arguments again. This will not recurse
-            // indefinitely because we don't do any of this extra logic if there was a separator.
-            vec_args.insert(index, "--".to_owned());
-            Self::try_parse_from(vec_args)
-        } else {
-            // If `index` were `None`, it would mean that all the args were environment variable
-            // declarations or part of the external args. But this cannot happen because the first
-            // argument should be the name of the binary being run and that is neither an
-            // environment variable declaration nor a external argument
-            unreachable!()
         }
     }
 
-    pub fn parse() -> Self {
-        match Self::try_parse_from(std::env::args()) {
-            Ok(options) => options,
-            Err(err) => {
-                eprintln!("{err}");
-                exit(1);
+    /// parse an iterator over command line arguments
+    pub fn try_parse_from<I, T>(iter: I) -> Result<Self, &'static str>
```

Comment:
Given that this function is now very generic and that we no longer have clap their test suite, I think this can do with a few unit tests showcasing proper handling of both common examples as well as some edge cases.